### PR TITLE
[Feature] Added ability to update and delete app folders

### LINKED ISSUE #1815: Add options to edit and delete folder
### Is your proposal related to a problem?
There is no options to edit and delete a folder

### Describe the solution you'd like
Create back-end APIs for 
 - Edit folder
 - Delete Folder
 
Front-end changes based on the design figma.com/file/kviZGq1lsML1Mi10OHYftU/1st-cycle-(Nov-jan)?node-id=505%3A8933

### LINKED ISSUE #3067: [feature] Organization level environment variables
### Is your proposal related to a problem?

Currently, we don't have an option to add organization level environment variables that any apps can access under an organization

### Describe the solution you'd like

We should be able to add, update, delete, and view environment variables.
- Should be able to add encrypted variables like a token.

### If the feature is approved, would you be willing to submit a PR?
Yes

### PR DESCRIPTION
**Pull request by shah21**
- [x] Delete & Update APIs and services
- [x] Delete & Update Group permissions
- [x] Implement folder menu & actions
- [x] API Specs

Resolves #1815

### CODE REVIEW

**Review by akshaysasidrn — CHANGES_REQUESTED** [review]

File: server/src/services/folders.service.ts [hunk 1]
```diff
@@ -292,4 +296,8 @@ export class FoldersService {
 
     return viewableAppsInFolder;
   }
+
+  async delete(id: string) {
+    return await this.foldersRepository.delete({ id });
```
Comment (akshaysasidrn) [inline server/src/services/folders.service.ts:301]: Please scope this with organization id

File: frontend/src/HomePage/Folders.jsx [hunk 2]
```diff
@@ -10,45 +13,147 @@ export const Folders = function Folders({
...
+
   return (
     <div className="w-100 px-3 pe-lg-4 folder-list">
+      <ConfirmDialog
+        show={showDeleteConfirmation}
+        message={`${
+          deletingFolder
+            ? `${
+                deletingFolder.folder_apps.length > 0
+                  ? `the folder ${deletingFolder.name} has ${deletingFolder.folder_apps.length} apps in it, by deleting the apps contain the folder will moved to uncategoried section.`
```
[the remark this answers is not in the document]
Comment (shah21) [reply frontend/src/HomePage/Folders.jsx:146]: Yeah, that will be fine, I just followed the design 👍

File: frontend/src/ManageGroupPermissionResources/ManageGroupPermissionResources.jsx [hunk 3]
```diff
@@ -614,6 +614,20 @@ class ManageGroupPermissionResources extends React.Component {
                                           />
                                           <span className="form-check-label">Create</span>
                                         </label>
+                                        <label className="form-check form-check-inline">
```
[the remark this answers is not in the document]
Comment (shah21) [reply frontend/src/ManageGroupPermissionResources/ManageGroupPermissionResources.jsx:617]: Yeah, that's right, I forgot to add the checkbox, I will add it

File: frontend/src/HomePage/FolderMenu.jsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import React from 'react';
+import OverlayTrigger from 'react-bootstrap/OverlayTrigger';
+import Popover from 'react-bootstrap/Popover';
+
+export const FolderMenu = function FolderMenu({
+  deleteFolder,
+  editFolder,
+  canDeleteApp = true,
```
[the remark this answers is not in the document]
Comment (shah21) [reply frontend/src/HomePage/FolderMenu.jsx:8]: That's a typo, I forgot to change the property's names, I should be like canDeleteFolder, canUpdateFolder, my bad

File: frontend/src/HomePage/Folders.jsx [hunk 5]
```diff
@@ -10,45 +13,140 @@ export const Folders = function Folders({
...
+      .deleteFolder(deletingFolder.id)
+      .then(() => {
+        toast.success('folder has been deleted.', {
+          position: 'top-center',
+        });
+        setShowDeleteConfirmation(false);
+        setDeletionStatus(false);
+        foldersChanged();
+      })
+      .catch(() => {
```
Comment (akshaysasidrn) [inline frontend/src/HomePage/Folders.jsx:96]: Please show an error toast as well

**Review by akshaysasidrn — CHANGES_REQUESTED** [review]

File: server/src/controllers/folders.controller.ts [hunk 6]
```diff
@@ -31,4 +32,30 @@ export class FoldersController {
     const folder = await this.foldersService.create(req.user, folderName);
     return decamelizeKeys(folder);
   }
+
+  @UseGuards(JwtAuthGuard)
+  @Put(':id')
+  async update(@Request() req, @Param('id') id, @Body() body: any) {
```
Comment (gsmithun4) [inline server/src/controllers/folders.controller.ts:38]:
```suggestion
  async update(@User() user, @Param('id') id, @Body() body: any) {
```

**Review by gsmithun4 — CHANGES_REQUESTED** [review]

File: server/src/controllers/folders.controller.ts [hunk 7]
```diff
@@ -31,4 +32,30 @@ export class FoldersController {
     const folder = await this.foldersService.create(req.user, folderName);
     return decamelizeKeys(folder);
   }
+
+  @UseGuards(JwtAuthGuard)
+  @Put(':id')
+  async update(@User() user, @Param('id') id, @Body() body: any) {
```
Comment (gsmithun4) [inline server/src/controllers/folders.controller.ts:38]:
```suggestion
  async update(@User() user, @Param('id') id, @Body('name') folderName: string) {
```

File: frontend/src/HomePage/Folders.jsx [hunk 8]
```diff
@@ -10,45 +13,142 @@ export const Folders = function Folders({
...
+    setShowUpdateForm(true);
+    setUpdatingFolder(folder);
+  }
+
+  function executeDeletion() {
+    setDeletionStatus(true);
+    folderService
+      .deleteFolder(deletingFolder.id)
+      .then(() => {
+        toast.success('folder has been deleted.', {
```
Comment (gsmithun4) [inline frontend/src/HomePage/Folders.jsx:89]: please capitalise all toasts

**Comment by Nandinisaha13** [comment]
<html>
<body>
<google-sheets-html-origin>Test Cases

Test Case | Steps | Result
-- | -- | --
Should be able to update folder name as an admin. | Login as admin. | Logged in as admin.
  | Create an application. | Application is created.
  | Create a folder | Folder is created.
  | Click on the 3 dots. | 2 options is displayed.
  | Click on edit folder. | The pop up is displayed.
  | Enter an updated name & click save. | The folder name is updated.
Should be able to delete folder as an admin. | Login as admin. | Logged in as admin.
  | Create an application. | Application is created.
  | Create a folder | Folder is created.
  | Click on the 3 dots. | 2 options is displayed.
  | Click on delete folder. | The pop up is displayed.
  | Click on yes. | The folder is deleted except the apps inside it.
Should be able to cancel the updated name of the folder. | Login as admin. | Logged in as admin.
  | Create an application. | Application is created.
  | Create a folder | Folder is created.
  | Click on the 3 dots. | 2 options is displayed.
  | Click on edit folder. | The pop up is displayed.
  | Click on cancel or cross. | The folder name is not updated.
Should be able to cancel the deletion of the folder. | Login as admin. | Logged in as admin.
  | Create an application. | Application is created.
  | Create a folder | Folder is created.
  | Click on the 3 dots. | 2 options is displayed.
  | Click on delete folder. | The pop up is displayed.
  | Click on cancel. | The folder is not deleted.
Should be able to update as non admin. | Login as admin. | Logged in as admin.
  | Create a new user. | User is created.
  | Create a new group. | New group is created.
  | Give folder create/update/delete permission along with application create permission. | The permissions are given.
  |   |  
  | Create an application. | Application is created.
  | Create a folder | Folder is created.
  | Click on the 3 dots. | 2 options is displayed.
  | Click on edit folder. | The pop up is displayed.
  | Enter an updated name & click save. | The folder name is updated.
Should be able to delete as non admin. | Login as admin. | Logged in as admin.
  | Create a new user. | User is created.
  | Create a new group. | New group is created.
  | Give folder create/update/delete permission along with application create permission. | The permissions are given.
  |   |  
  | Create an application. | Application is created.
  | Create a folder | Folder is created.
  | Click on the 3 dots. | 2 options is displayed.
  | Click on delete folder. | The pop up is displayed.
  | Click on yes. | The folder is deleted except the apps inside it.


</body>
</html>

**Review by akshaysasidrn — APPROVED** [review]

**Pull request merged**